lib/sqlite3/database.rb

```diff
@@ -691,6 +691,21 @@ def readonly?
       @readonly
     end
 
+    # Sets a #busy_handler that releases the GVL between retries,
+    # but only retries up to the indicated number of +milliseconds+.
+    # This is an alternative to #busy_timeout, which holds the GVL
+    # while SQLite sleeps and retries.
+    def busy_handler_timeout=( milliseconds )
+      timeout_seconds = milliseconds.fdiv(1000)
+      timeout_deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout_seconds
```

Comment:
@fractaledmind Looking at the code again, I believe there is a bug here.   

This is setting a clock based timeout deadline at the time `busy_handler_timeout` is set rather than when the busy_handler gets invoked for the first time.   Essentially setting a hard timeout for `timeout_seconds` for all invocations of the busy handler.

Consider the following (I didn't test it but I believe this should show the issue, if I am reading the code correctly)

```
busy_handler_timeout_db = SQLite3::Database.open( "test.db" )
busy_handler_timeout_db.busy_handler_timeout = 1000
sleep 1.5

# timeout_deadline should be expired now since that was set 
# at the point in time when we set busy_handler_timeout to 1000.
# Any invocations of the busy handler will now throw busy errors
```

Reply:
I think you're right. Great catch. Will fix when I'm back at my laptop

Reply:
@suwyn Checkout https://github.com/sparklemotion/sqlite3-ruby/pull/443/commits/48daae13fb32213c84a07ea7e7ed8714d0462fb6

Reply:
Looks better @fractaledmind 

It still has me wondering if it's thread safe though since it is using an instance variable.   If two threads used the same connection, you'd have an issue.   I did some experimenting in Rails and the connection pool appears to be thread safe (as it states in the docs) .

Experimenting with the gem on its own, I wasn't able quickly able to cause any issue.   SQLite doesnt have a way to sleep so I stopped short of writing a long running UDF.  

I think the ideal solution is to either bake the busy handler into the C code or pass the time of first invocation into the handler from C, like it does with count.   I'd obviously strongly prefer the former, but have no idea on what that would take and outside of the scope of this PR.

So short of using the count to estimate the clock time (e.g. the `(count * RETRY_INTERVAL) > timeout_seconds` approach) instead of the clock time we'll just have to know that a race condition probably exists.

Reply:
@suwyn: You should join this Discord server where we talk a lot about the SQLite ecosystem in Ruby: https://discord.gg/ehdDh5C4

That would allow us to connect more easily and even pair on this. I'd love some help on writing useful tests. I have a long running UDF already written:
```sql
WITH RECURSIVE r(i) AS (
  VALUES(0)
  UNION ALL
  SELECT i FROM r
  LIMIT 10000000
)
SELECT i FROM r ORDER BY i LIMIT 1;
```
What I don't have are resilient and expansive tests. Want to work on getting this over the line together?